US 17 creation page contact

Samaze123 opened the pull request:
# US17

US perso de Samuel Lambert.
Creation de la page contact. Cette page est necessaire pour que l'utilisateur qui le souhaite puisse prendre contact avec l'equipe de la sonde.
Lors du contact avec l'equipe, le message de l'utilisateur est sauve dans un ticket.
Pour que l'equipe puisse lire et manager ces tickets, il a fallu creer une page, uniquement pour l'equipe.
Ensuite, pour plus de classification de ces tickets j'ai cree les categories qui sont manager exclusivement par l'equipe mais qui sont utilisable par l'utilisateur.

## BACKEND

### Modifications Backend

#### Routes

- index.ts : Ajout de toutes les routes dans ce fichier pour facilite l'import
user.js : Ajout d'une route pour recevoir l'ability d'un utilisateur.
- auth.js : Ajout d'une reponse avec des donnees de l'utilisateur qui seront stockee dans le localstorage du front. Ajout de ability lors de la creation d'un compte.
     
- category.js : Creation de la route pour ajouter, supprimer et recevoir les categories.
  - Il y a aussi une route qui sert a recevoir seulement une categorie grace a son ID.

- ticket.js : Creation de la route pour ajouter, supprimer et recevoir les tickets.
  - Utilisation du middleware RequireAdmin utilise pou certaines url.
  - Details lors du retour si mauvaises donnees entrees. 
  - Il y a aussi une route qui sert a recevoir seulement un ticket grace a son ID.

#### Middleware

- index.ts : Ajout de tous les middleware dans ce fichier pour facilite l'import.
- auth.js : Ajout du middleware requireAdmin qui verifie que l'utilisateur est connecte et possede les bons droits admin.

#### Modeles

- index.ts : Ajout de tous les modeles dans ce fichier pour facilite l'import.
- tickets.js : Ajout du modele Ticket qui correspond a celui de la base de donnees.
- category.js : Creation du modele Category
- user.js : Ajout de l'attribut ability dans le User ce qui permet d'etre admin des tickets et d'avoir les droits de :
  - visionner les tickets ouverts.
  - supprimer les tickets.

#### Autre

- app.js : Ajout de la route contact et category dans le routeur backend.
  
### Tests Backend
- cypress.config.ts : Ajout d'un parametre pour que les tests prennent des videos.
- ticket.test.js : Creation de test pour la route /api/tickets.
  - Il y a aussi des petits tests en rapport avec les categories.
  Pour effectuer les tests backend, il faut utiliser la commande : npm run test

## FRONTEND

### Modification Frontend

#### Composants

- index.ts : Ajout de tous les composants dans ce fichier pour facilite l'import.
- Ticket.tsx : Creation du composant Ticket qui permet de l'utiliser plus facilement plusieurs fois.
- Navbar.tsx : Ajout d'un bouton pour acceder a la nouvelle page contact.
- UserProvider.tsx : Ajout du provider pour concentrer toutes la variables "globales" (et inscrites dans le local storage) du site.
- UserContext.tsx : Ajout du UserContext qui permet d'acceder au UserProvider.
- Navbar.tsx : Ajout de liens vers les nouvelles pages et changement de navbar en fonction des droits et de l'etat de l'utilisateur.
- components/Popup.tsx : Ajout d'un popup pour afficher a l'utilisateur qu'il a bien envoye son ticket
- CategoryCard.tsx : Ajout de la carte Categorie pour plus de facilite lors de son utilisation multiple.

#### Pages

- index.ts : Ajout de toutes les pages dans ce fichier pour facilite l'import.
- About.tsx : Ajout de la navbar dans la page.
- CategoryManager : Creation de la page admin pour gerer les categories.
  - La page ne s'affiche seulement lorsque les droits de l'utilisateur ont ete checke auparavant.
- ContactUs.tsx : Creation de la page contact.
  - Condition avec les inputs et toast error si ces tests ne sont pas valides.
  - Pour acceder a cette page, il faut cliquer sur la bouton dans la navbar ou aller sur LeNomDuSite/contact.
- Dashboard.tsx : Ajout d'un lien vers la page contact car pas de navbar utilisable.
- Login.tsx : Ajout de la reponse lors du login dans le local storage pour sauvegarder certains donnees insensibles. (Les donnees sauvegardee dans le local storage ne sont pas utilisees pour les requetes mais seulement afficher plus facilement certaines parties du frontend)
  - Ajout d'un toast pour signaler une erreur lors d'une redirection forcee a l'utilisateur.
- TicketsViewer.tsx : Creation de la page admin pour manager les tickets (suppression, archivage et prise de contact avec l'utilisateur par mail)
  - La page ne s'affiche seulement si les droits ont ete checke auparavant.
  - Lors de la suppression ou le changement d'etat d'un ticket, la page est mise a jour sans etre rafraichie.
  - Details lors de la suppression d'un ticket.
  - Pas de pagination pour cette page. S'il y a beaucoup de tickets a charger d'un seul coup, peut poser problemes.
  - Bouton qui redirige vers le manager des categories exlusivement reserve aussi aux admins.

#### Types

Ce dosser contient toutes les interfaces utilisees par certains composants et variables
- index.ts : Ajout de tous les types dans ce fichier pour facilite l'import.

#### Autres

- app.tsx : Ajout de la page Contact, TicketViewer et CategoryManager dans le router du site.
- main.tsx : Ajout du UserProvider dans l'application pour le mettre accessible a toutes la pages.

### Tests Frontend

#### Tests End-To-End

- contact.cy.ts : Contient tous les tests e2e pour la page contact
  - Soumission du formulaire avec un probleme dans le nom de l'utilisateur.
  - Soumission du formulaire sans erreurs.
  - Soumission du formulaire avec des cases vides.

Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.e2e

#### Tests unitaires

- App.test.tsx : Ajout du test de l'app dans le dossier qui sera utilise lors de la commande.

Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.unit

JarodEphec reviewed (approved):
Ici, la pr a déjà été validé dans la pr #22 , nous avons eu des soucis avec Github lors du merge ce qui nous a obligés à revert le merge,  dans ce procédé, Samuel a eu des problèmes avec son code et a donc dû le retravailler. J'ai tout de même tout retester, tout fonctionne comme les tests qui passent tous. Bon travail !

Samaze123 committed:
# US17

US perso de Samuel Lambert.
Creation de la page contact. Cette page est necessaire pour que l'utilisateur qui le souhaite puisse prendre contact avec l'equipe de la sonde.
Lors du contact avec l'equipe, le message de l'utilisateur est sauve dans un ticket.
Pour que l'equipe puisse lire et manager ces tickets, il a fallu creer une page, uniquement pour l'equipe.
Ensuite, pour plus de classification de ces tickets j'ai cree les categories qui sont manager exclusivement par l'equipe mais qui sont utilisable par l'utilisateur.

## BACKEND

### Modifications Backend

#### Routes

- index.ts : Ajout de toutes les routes dans ce fichier pour facilite l'import
user.js : Ajout d'une route pour recevoir l'ability d'un utilisateur.
- auth.js : Ajout d'une reponse avec des donnees de l'utilisateur qui seront stockee dans le localstorage du front. Ajout de ability lors de la creation d'un compte.

- category.js : Creation de la route pour ajouter, supprimer et recevoir les categories.
  - Il y a aussi une route qui sert a recevoir seulement une categorie grace a son ID.

- ticket.js : Creation de la route pour ajouter, supprimer et recevoir les tickets.
  - Utilisation du middleware RequireAdmin utilise pou certaines url.
  - Details lors du retour si mauvaises donnees entrees.
  - Il y a aussi une route qui sert a recevoir seulement un ticket grace a son ID.

#### Middleware

- index.ts : Ajout de tous les middleware dans ce fichier pour facilite l'import.
- auth.js : Ajout du middleware requireAdmin qui verifie que l'utilisateur est connecte et possede les bons droits admin.

#### Modeles

- index.ts : Ajout de tous les modeles dans ce fichier pour facilite l'import.
- tickets.js : Ajout du modele Ticket qui correspond a celui de la base de donnees.
- category.js : Creation du modele Category
- user.js : Ajout de l'attribut ability dans le User ce qui permet d'etre admin des tickets et d'avoir les droits de :
  - visionner les tickets ouverts.
  - supprimer les tickets.

#### Autre

- app.js : Ajout de la route contact et category dans le routeur backend.

### Tests Backend
- cypress.config.ts : Ajout d'un parametre pour que les tests prennent des videos.
- ticket.test.js : Creation de test pour la route /api/tickets.
  - Il y a aussi des petits tests en rapport avec les categories.
  Pour effectuer les tests backend, il faut utiliser la commande : npm run test

## FRONTEND

### Modification Frontend

#### Composants

- index.ts : Ajout de tous les composants dans ce fichier pour facilite l'import.
- Ticket.tsx : Creation du composant Ticket qui permet de l'utiliser plus facilement plusieurs fois.
- Navbar.tsx : Ajout d'un bouton pour acceder a la nouvelle page contact.
- UserProvider.tsx : Ajout du provider pour concentrer toutes la variables "globales" (et inscrites dans le local storage) du site.
- UserContext.tsx : Ajout du UserContext qui permet d'acceder au UserProvider.
- Navbar.tsx : Ajout de liens vers les nouvelles pages et changement de navbar en fonction des droits et de l'etat de l'utilisateur.
- components/Popup.tsx : Ajout d'un popup pour afficher a l'utilisateur qu'il a bien envoye son ticket
- CategoryCard.tsx : Ajout de la carte Categorie pour plus de facilite lors de son utilisation multiple.

#### Pages

- index.ts : Ajout de toutes les pages dans ce fichier pour facilite l'import.
- About.tsx : Ajout de la navbar dans la page.
- CategoryManager : Creation de la page admin pour gerer les categories.
  - La page ne s'affiche seulement lorsque les droits de l'utilisateur ont ete checke auparavant.
- ContactUs.tsx : Creation de la page contact.
  - Condition avec les inputs et toast error si ces tests ne sont pas valides.
  - Pour acceder a cette page, il faut cliquer sur la bouton dans la navbar ou aller sur LeNomDuSite/contact.
- Dashboard.tsx : Ajout d'un lien vers la page contact car pas de navbar utilisable.
- Login.tsx : Ajout de la reponse lors du login dans le local storage pour sauvegarder certains donnees insensibles. (Les donnees sauvegardee dans le local storage ne sont pas utilisees pour les requetes mais seulement afficher plus facilement certaines parties du frontend)
  - Ajout d'un toast pour signaler une erreur lors d'une redirection forcee a l'utilisateur.
- TicketsViewer.tsx : Creation de la page admin pour manager les tickets (suppression, archivage et prise de contact avec l'utilisateur par mail)
  - La page ne s'affiche seulement si les droits ont ete checke auparavant.
  - Lors de la suppression ou le changement d'etat d'un ticket, la page est mise a jour sans etre rafraichie.
  - Details lors de la suppression d'un ticket.
  - Pas de pagination pour cette page. S'il y a beaucoup de tickets a charger d'un seul coup, peut poser problemes.
  - Bouton qui redirige vers le manager des categories exlusivement reserve aussi aux admins.

#### Types

Ce dosser contient toutes les interfaces utilisees par certains composants et variables
- index.ts : Ajout de tous les types dans ce fichier pour facilite l'import.

#### Autres

- app.tsx : Ajout de la page Contact, TicketViewer et CategoryManager dans le router du site.
- main.tsx : Ajout du UserProvider dans l'application pour le mettre accessible a toutes la pages.

### Tests Frontend

#### Tests End-To-End

- contact.cy.ts : Contient tous les tests e2e pour la page contact
  - Soumission du formulaire avec un probleme dans le nom de l'utilisateur.
  - Soumission du formulaire sans erreurs.
  - Soumission du formulaire avec des cases vides.

Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.e2e

#### Tests unitaires

- App.test.tsx : Ajout du test de l'app dans le dossier qui sera utilise lors de la commande.

Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.unit
